src/hashtable.c

```diff
@@ -2025,8 +2031,17 @@ void hashtableReleaseIterator(hashtableIterator *iterator) {
  * Returns false if there are no more entries. */
 bool hashtableNext(hashtableIterator *iterator, void **elemptr) {
     iter *iter = iteratorFromOpaque(iterator);
+
+    assert(iter->table <= HASHTABLE_ITER_FINISHED);
+    /* Check if iterator is exhausted or in undefined state */
+    if (iter->table == HASHTABLE_ITER_FINISHED ||
+        (iter->table == HASHTABLE_ITER_REHASH_TABLE && !hashtableIsRehashing(iter->hashtable)) ||
+        (iter->index >= 0 && iter->hashtable->tables[iter->table] && (size_t)iter->index >= numBuckets(iter->hashtable->bucket_exp[iter->table]))) {
```

Comment:
I still think we should separate the 2 cases. the first one (repeat calling next after we completed the iteration) is not expected, but is now considered "legal", the second case is "unexpected" and usually will indicate some bug.
```suggestion
    serverDebugAssert((iter->table == HASHTABLE_ITER_REHASH_TABLE && !hashtableIsRehashing(iter->hashtable)) ||
        (iter->index >= 0 && iter->hashtable->tables[iter->table] && (size_t)iter->index >= numBuckets(iter->hashtable->bucket_exp[iter->table])));
    if (iter->table == HASHTABLE_ITER_FINISHED) {
```

Reply:
Done